python/cog/predictor.py

```diff
@@ -99,18 +101,34 @@ def run_setup(predictor: BasePredictor) -> None:
     # up a little bit.
     # TODO: CogFile/CogPath should have subclasses for each of the subtypes
     if weights_url:
-        if weights_type == CogFile:
-            weights = cast(CogFile, CogFile.validate(weights_url))
-        elif weights_type == CogPath:
-            # TODO: So this can be a url. evil!
-            weights = cast(CogPath, CogPath.validate(weights_url))
-        # allow people to download weights themselves
-        elif weights_type == str:  # noqa: E721
-            weights = weights_url
+        if PYDANTIC_V2:
+            from pydantic import TypeAdapter
+
+            for t in [CogFile, CogPath]:
+                try:
+                    weights = TypeAdapter(t).validate_python(weights_url)
+                    break
+                except Exception:  # pylint: disable=broad-except # noqa: S110
+                    pass
```

Comment:
If none of these work do we need to raise an exception?

Reply:
That should be handled by the `else` clause below